### PR TITLE
Feature 12 resource availability

### PR DESCRIPTION
**Pull request by Alekseeybg**

### CODE REVIEW

File: src/main/java/jewellery/inventory/controller/ResourceAvailabilityController.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+@RestController
+@RequestMapping("/resources/availability")
+@CrossOrigin(origins = "${cors.origins}")
+@RequiredArgsConstructor
+public class ResourceAvailabilityController {
+  private final ResourceAvailabilityService resourceAvailabilityService;
+
+  @ResponseStatus(HttpStatus.CREATED)
+  @PostMapping
+  public UserResponseDto addResourceToUser(
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/controller/ResourceAvailabilityController.java]: Let's use the ResourceInUserResponseDto instead of the user one.

File: src/main/java/jewellery/inventory/dto/request/resource/ResourceInUserRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+@Data
+public class ResourceInUserRequestDto {
+  private static final String QUANTITY_MIN_VALUE_MSG = "Quantity should not be less than 1.";
+  private static final String QUANTITY_DECIMAL_PLACES_MSG =
+      "Quantity should not have more than 2 decimal places.";
+
+  @NotNull private UUID userId;
+  @NotNull private UUID resourceId;
+
+  @Min(value = 1, message = QUANTITY_MIN_VALUE_MSG)
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/dto/request/resource/ResourceInUserRequestDto.java]: quantity could be less than 1. For example, we could want to add 0.5g of gold.

File: src/main/java/jewellery/inventory/dto/response/UserResponseDto.java [hunk 3]
```diff
@@ -1,11 +1,14 @@
...
 import java.util.UUID;
+import jewellery.inventory.dto.response.resource.ResourceInUserResponseDto;
 import lombok.Data;
 
 @Data
 public class UserResponseDto {
   private UUID id;
   private String name;
   private String email;
+  List<ResourceInUserResponseDto> resources;
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/dto/response/UserResponseDto.java]: Let's not update the UserResponseDto but modify the ResourceInUserResponseDto to have the UserResponseDto as a field and also a list of ResourceResponseDtos and quantities

File: src/main/java/jewellery/inventory/dto/response/resource/ResourceInUserResponseDto.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+import java.util.UUID;
+import lombok.Data;
+
+@Data
+public class ResourceInUserResponseDto {
+  private UUID resourceId;
+  private String ownerName;
+  private String resourceClazz;
+  private double quantity;
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/dto/response/resource/ResourceInUserResponseDto.java]: Can we rework this DTO as mentioned above to have the UserDto as a field and a list of ResourceResponseDtos with quantities. 
 
For the GET /availability/{userId}/{resourceId} we can create a SingleResourceInUserResponseDto that has only one ResourceResponseDto instead of a list. (tbh, I am currently wondering whether we would even need the singular resourceInUser endpoint at all)

File: src/main/java/jewellery/inventory/exception/GlobalExceptionHandler.java [hunk 5]
```diff
@@ -43,6 +45,18 @@ public ResponseEntity<Object> handleDuplicateException(DuplicateException ex) {
...
+  public ResponseEntity<Object> handleInvalidResourceQuantityException(
+      InvalidResourceQuantityException ex) {
+    return createErrorResponse(HttpStatus.BAD_REQUEST, ex.getMessage());
+  }
+
+  @ExceptionHandler(UserDoesNotOwnResourceException.class)
+  public ResponseEntity<Object> handleUserDoesNotOwnResourceException(
+      UserDoesNotOwnResourceException ex) {
+    return createErrorResponse(HttpStatus.BAD_REQUEST, ex.getMessage());
+  }
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/exception/GlobalExceptionHandler.java]: I notice these methods are pretty much the same. If we have the same error handling for multiple exceptions we can group them in a single method according to https://stackoverflow.com/a/33552395/13062206

File: src/main/java/jewellery/inventory/exception/duplicateException/DuplicateEmailException.java [hunk 6]
```diff
@@ -1,4 +1,4 @@
-package jewellery.inventory.exception;
+package jewellery.inventory.exception.duplicateException;
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/exception/duplicateException/DuplicateEmailException.java]: package names should all be lower cased or snake_cased -> https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html 
Also, given the classes are already in an `exception` package, it doesn't make much sense to include it in the name of the inner package (duplicateException). I would name it just `jewellery.inventory.exception.duplicate`

File: src/main/java/jewellery/inventory/exception/invalidResourceQuantityException/InsufficientResourceQuantityException.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package jewellery.inventory.exception.invalidResourceQuantityException;
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/exception/invalidResourceQuantityException/InsufficientResourceQuantityException.java]: same here, rename it to either `invalidresourcequantity` or `invalid_resource_quantity`

File: src/main/java/jewellery/inventory/exception/notFoundException/NotFoundException.java [hunk 8]
```diff
@@ -1,4 +1,4 @@
-package jewellery.inventory.exception;
+package jewellery.inventory.exception.notFoundException;
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/exception/notFoundException/NotFoundException.java]: same here -> `notfound` or `not_found`

File: src/main/java/jewellery/inventory/model/User.java [hunk 9]
```diff
@@ -21,8 +22,14 @@ public class User {
...
-  private List<Product> productsOwned;
+  private List<Product> productsOwned = new ArrayList<>();
+
+  @OneToMany(mappedBy = "owner", cascade = CascadeType.ALL, orphanRemoval = true)
+  private List<ResourceInUser> resourcesOwned = new ArrayList<>();
+
+  public void addResource(ResourceInUser resourceInUser) {
+    resourcesOwned.add(resourceInUser);
+    resourceInUser.setOwner(this);
+  }
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/model/User.java]: This method is only used in tests from what I saw. Please, move it to the test package in this case. We shouldn't expose methods for global use if we are only using them in the tests.

File: src/main/java/jewellery/inventory/service/ResourceAvailabilityService.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ResourceAvailabilityService.java:1]: We are now using a double naming convention (I take the blame for it) -> ResourceAvailability and ResourceInUser. Let's just use ResourceInUser everywhere as I think it's more descriptive of what the object/service/repo is doing. We can just keep the endpoint as `resources/availability`.

File: src/main/java/jewellery/inventory/service/ResourceAvailabilityService.java [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    resourceInUser.setQuantity(resourceInUser.getQuantity() + resourceUserDto.getQuantity());
+
+    return UserMapper.INSTANCE.toUserResponse(userRepository.save(user));
+  }
+
+  public double getUserResourceQuantity(UUID userId, UUID resourceId) {
+    User user = findUserById(userId);
+    ResourceInUser resourceInUser =
+        findResourceInUser(user, resourceId)
+            .orElseThrow(() -> new ResourceNotFoundException(resourceId));
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ResourceAvailabilityService.java]: Here we are throwing a `ResourceNotFoundException`. This is a bit misleading as this will be thrown if the resource is not found in the user's possessions.
Let's create a `ResourceInUserNotFoundException` and use it instead.

File: src/test/java/jewellery/inventory/helper/UserTestHelper.java [hunk 11]
```diff
@@ -56,31 +54,40 @@ public static UserRequestDto createDifferentUserRequest() {
...
+    gemstoneRequestDto.setQuantityType("5");
+    gemstoneRequestDto.setColor("blue");
+    gemstoneRequestDto.setCarat(15.0);
+    gemstoneRequestDto.setCut("round");
+    gemstoneRequestDto.setClarity("VS1");
+    gemstoneRequestDto.setDimensionX(1.0);
+    gemstoneRequestDto.setDimensionY(1.0);
+    gemstoneRequestDto.setDimensionZ(1.0);
+    gemstoneRequestDto.setShape("oval");
+    return gemstoneRequestDto;
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/helper/UserTestHelper.java]: There is a GemstoneRequestDto creation method in ResourceTestHelper -> getGemstoneRequestDto. Please, use it instead

File: src/main/java/jewellery/inventory/controller/ResourceAvailabilityController.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+      @RequestBody @Valid ResourceInUserRequestDto resourceUserDto) {
+    return resourceAvailabilityService.addResourceToUser(resourceUserDto);
+  }
+
+  @ResponseStatus(HttpStatus.OK)
+  @GetMapping("/{userId}/{resourceId}")
+  public double getUserResourceQuantity(@PathVariable UUID userId, @PathVariable UUID resourceId) {
+    return resourceAvailabilityService.getUserResourceQuantity(userId, resourceId);
+  }
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/controller/ResourceAvailabilityController.java]: After some thought, I don't think we need this endpoint at all. Can you please, remove it and all the related methods/dtos.

File: src/main/java/jewellery/inventory/service/ResourceAvailabilityService.java [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+    double totalQuantity = resourceInUser.getQuantity();
+    double newQuantity = totalQuantity - quantity;
+    if (newQuantity < 0) {
+      throw new InsufficientResourceQuantityException(quantity, totalQuantity);
+    }
+
+    resourceInUser.setQuantity(newQuantity);
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ResourceAvailabilityService.java]: Let's make it so that if we are removing all of the quantity, then we also remove the resource from the user.

File: src/test/java/jewellery/inventory/integration/ResourceAvailabilityCrudIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,422 @@
...
+        this.testRestTemplate.exchange(
+            url, HttpMethod.GET, null, new ParameterizedTypeReference<>() {});
+
+    assertEquals(HttpStatus.OK, response.getStatusCode());
+    assertNotNull(response.getBody());
+    assertTrue(response.getBody().isEmpty());
+  }
+
+  @Test
+  void willReturnNotFoundWhenGettingResourcesFromNonExistingUser() {
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/ResourceAvailabilityCrudIntegrationTest.java]: This seems to be duplicate of `willReturnNotFoundWhenUserDoesNotExist`

File: src/test/java/jewellery/inventory/integration/ResourceAvailabilityCrudIntegrationTest.java [hunk 15]
```diff
@@ -0,0 +1,422 @@
...
+    String url = getResourceAvailabilityUrl(createdUser.getId(), nonExistentResourceId) + "/" + 1;
+
+    ResponseEntity<String> response =
+        testRestTemplate.exchange(url, HttpMethod.DELETE, HttpEntity.EMPTY, String.class);
+
+    assertEquals(HttpStatus.NOT_FOUND, response.getStatusCode());
+  }
+
+  @Test
+  void willThrowUserNotExistsExceptionWhenUserDoesNotExist() {
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/ResourceAvailabilityCrudIntegrationTest.java]: This method doesn't indicate it's trying to delete from non existing user. It looks like other methods (GET, PUT) that are dealing with non existing user.

File: src/test/java/jewellery/inventory/integration/ResourceAvailabilityCrudIntegrationTest.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/ResourceAvailabilityCrudIntegrationTest.java:1]: We should think of some naming convention for the tests to indicate what is the difference between them. We are testing a lot of negative scenarios that are very similar to one another - e.g. 1. removing a resource from a user fails for reason x, 2. removing quantity of a resource from a user fails for reason x, 3.fetching resources from a user fails for reason x. All of the mentioned are failing for the same reason but we are testing them for all of the different endpoints. If it were me, I wouldn't have tested all of them but I think I am fine having all of them for now. Right now it's difficult to understand what the methods are trying to do, though. A lot of them are just telling us the outcome of the test, not what caused it. 

TLDR: let's try to think of a naming convention for the tests:
See if this works:
`<action>ResourcesInUser<Sucess/Fail>When<reason>`
examples:
fetchResourcesInUserSuccessfully
fetchResourcesInUserFailsWhenUserNotFound
removeResourcesInUserFailsWhenResourceInUserNotFound

File: src/test/java/jewellery/inventory/unit/service/ResourceAvailabilityServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,378 @@
...
+    verify(userRepository, never()).save(any(User.class));
+  }
+
+  @Test
+  @DisplayName("Should throw ResourceNotFoundException when resource does not exist")
+  void addResourceToUserWhenResourceDoesNotExistThenThrowException() {
+    ResourceInUserRequestDto resourceUserDto =
+        createResourceInUserRequestDto(userId, resourceId, 10);
+
+    when(userRepository.findById(userId)).thenReturn(Optional.of(new User())); // changed this line
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/unit/service/ResourceAvailabilityServiceTest.java]: comment doesn't add any value. please, remove it. if you think there is any template-like/copy-paste between methods, please extract them in methods and use them.

File: src/test/java/jewellery/inventory/unit/service/ResourceAvailabilityServiceTest.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/unit/service/ResourceAvailabilityServiceTest.java:1]: A lot of these tests are covered in the integration tests I think. We should think of a testing strategy because it doesn't make sense to test the same behaviour twice.

**Review by VladoKat — CHANGES_REQUESTED** [review]
There are some changes that need to be made. They are described in the comments. 
Our tests are way too long now. This will make them unmanageable at some point. I will research a bit to see what testing strategy is best to implement but I think the rule of thumb should be:
Have integration tests for the scenarios we think are going to be most commonly used. This may include some negative scenarios but not all of them.
The corner cases and specific error cases scenarios should be tested in service tests.

Let me double check the rules above and we may discuss after how to change (if at all) our testing strategy.

File: src/main/java/jewellery/inventory/dto/response/resource/ResourceInUserResponseDto.java (hunk 4 above)
Comment (VladoKat) [reply src/main/java/jewellery/inventory/dto/response/resource/ResourceInUserResponseDto.java]: Just to make sure we avoid confusion, we are not going to have a SingleResourceInUserResponseDto as it is mentioned in a comment below.

File: src/main/java/jewellery/inventory/dto/request/resource/ResourceInUserRequestDto.java (hunk 2 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/dto/request/resource/ResourceInUserRequestDto.java]: Fixed

File: src/main/java/jewellery/inventory/exception/notFoundException/NotFoundException.java (hunk 8 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/exception/notFoundException/NotFoundException.java]: Fixed

File: src/main/java/jewellery/inventory/exception/invalidResourceQuantityException/InsufficientResourceQuantityException.java (hunk 7 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/exception/invalidResourceQuantityException/InsufficientResourceQuantityException.java]: Fixed

File: src/main/java/jewellery/inventory/exception/duplicateException/DuplicateEmailException.java (hunk 6 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/exception/duplicateException/DuplicateEmailException.java]: While I was renaming them I felt how wrong I did it. Was thinking about asking about it or research it but never managed to do that.

File: src/test/java/jewellery/inventory/helper/UserTestHelper.java (hunk 11 above)
Comment (Alekseeybg) [reply src/test/java/jewellery/inventory/helper/UserTestHelper.java]: Fixed

File: src/test/java/jewellery/inventory/integration/ResourceAvailabilityCrudIntegrationTest.java (hunk 14 above)
Comment (Alekseeybg) [reply src/test/java/jewellery/inventory/integration/ResourceAvailabilityCrudIntegrationTest.java]: Fixed

File: src/main/java/jewellery/inventory/controller/ResourceAvailabilityController.java (hunk 12 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/controller/ResourceAvailabilityController.java]: Removed

File: src/main/java/jewellery/inventory/service/ResourceAvailabilityService.java (hunk 10 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/service/ResourceAvailabilityService.java]: Fixed

File: src/main/java/jewellery/inventory/controller/ResourceAvailabilityController.java (hunk 1 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/controller/ResourceAvailabilityController.java]: Fixed

File: src/main/java/jewellery/inventory/dto/response/UserResponseDto.java (hunk 3 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/dto/response/UserResponseDto.java]: Fixed

File: src/main/java/jewellery/inventory/dto/response/resource/ResourceInUserResponseDto.java (hunk 4 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/dto/response/resource/ResourceInUserResponseDto.java]: Done

File: src/main/java/jewellery/inventory/exception/GlobalExceptionHandler.java (hunk 5 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/exception/GlobalExceptionHandler.java]: Fixed

File: src/main/java/jewellery/inventory/service/ResourceAvailabilityService.java
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/service/ResourceAvailabilityService.java:1]: Fixed

File: src/test/java/jewellery/inventory/unit/service/ResourceAvailabilityServiceTest.java (hunk 16 above)
Comment (Alekseeybg) [reply src/test/java/jewellery/inventory/unit/service/ResourceAvailabilityServiceTest.java]: Removed

File: src/main/java/jewellery/inventory/model/User.java (hunk 9 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/model/User.java]: Fixed

File: src/test/java/jewellery/inventory/integration/ResourceAvailabilityCrudIntegrationTest.java (hunk 15 above)
Comment (Alekseeybg) [reply src/test/java/jewellery/inventory/integration/ResourceAvailabilityCrudIntegrationTest.java]: Fixed

File: src/test/java/jewellery/inventory/integration/ResourceAvailabilityCrudIntegrationTest.java
Comment (Alekseeybg) [reply src/test/java/jewellery/inventory/integration/ResourceAvailabilityCrudIntegrationTest.java:1]: Fixed all integration test names as suggested. Both for ResourceInUser and User integration tests

File: src/main/java/jewellery/inventory/service/ResourceAvailabilityService.java (hunk 13 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/service/ResourceAvailabilityService.java]: Fixed

File: src/main/java/jewellery/inventory/exception/GlobalExceptionHandler.java [hunk 17]
```diff
@@ -45,15 +45,9 @@ public ResponseEntity<Object> handleDuplicateException(DuplicateException ex) {
...
-  @ExceptionHandler(InvalidResourceQuantityException.class)
-  public ResponseEntity<Object> handleInvalidResourceQuantityException(
-      InvalidResourceQuantityException ex) {
-    return createErrorResponse(HttpStatus.BAD_REQUEST, ex.getMessage());
-  }
-
-  @ExceptionHandler(UserDoesNotOwnResourceException.class)
-  public ResponseEntity<Object> handleUserDoesNotOwnResourceException(
-      UserDoesNotOwnResourceException ex) {
+  @ExceptionHandler({InvalidResourceQuantityException.class, UserDoesNotOwnResourceException.class})
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/exception/GlobalExceptionHandler.java]: Is `UserDoesNotOwnResourceException` a bad data exception? I assume it's thrown when we are trying to update/delete ResourceInUser that the user doesn't posses. In this case, shouldn't it be a not found?

File: src/main/java/jewellery/inventory/exception/GlobalExceptionHandler.java [hunk 18]
```diff
@@ -45,15 +45,9 @@ public ResponseEntity<Object> handleDuplicateException(DuplicateException ex) {
...
+  public ResponseEntity<Object> handleResourceExceptions(
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/exception/GlobalExceptionHandler.java]: The name suggests we are handling only resourceExceptions. Let's name it a bit more generic as it is handling all exceptions resulting in BAD_DATA response code: handleBadDataExceptions maybe?

File: src/main/java/jewellery/inventory/exception/GlobalExceptionHandler.java (hunk 17 above)
Comment (VladoKat) [inline src/main/java/jewellery/inventory/exception/GlobalExceptionHandler.java]: Should we include the DuplicateException here as well? I see the `handleDuplicateException` function does the same.

File: src/main/java/jewellery/inventory/exception/not_found/ResourceInUserNotFoundException.java [hunk 19]
```diff
@@ -0,0 +1,9 @@
+package jewellery.inventory.exception.not_found;
+
+import java.util.UUID;
+
+public class ResourceInUserNotFoundException extends NotFoundException {
+    public ResourceInUserNotFoundException(UUID id) {
+        super("The resource with id " + id + " is not owned by this user");
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/exception/not_found/ResourceInUserNotFoundException.java]: Could we also include the user UUID in this exception?

File: src/main/java/jewellery/inventory/mapper/UserMapper.java [hunk 20]
```diff
@@ -26,19 +27,23 @@ public interface UserMapper {
...
-        resourcesDtoList.add(toResourceInUserResponseDto(resourceInUser));
-      }
-      userResponseDto.setResources(resourcesDtoList);
+  default void enrichResourceInUserResponseDto(
+      ResourceInUser resourceInUser,
+      @MappingTarget ResourceInUserResponseDto resourceInUserResponseDto) {
+    if (resourceInUser.getResource() != null) {
+      ResourceResponseDto resourceResponseDto =
+          ResourceMapper.toResourceResponse(resourceInUser.getResource());
+      resourceInUserResponseDto.setResources(Collections.singletonList(resourceResponseDto));
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/mapper/UserMapper.java]: I might be getting this wrong but it seems to me the ResourceInUserResponseDto will always have just the one resource inside it. It could have multiple

File: src/main/java/jewellery/inventory/service/ResourceInUserService.java [hunk 21]
```diff
@@ -73,6 +68,9 @@ public void removeQuantityFromResource(UUID userId, UUID resourceId, double quan
     }
 
     resourceInUser.setQuantity(newQuantity);
+    if (newQuantity == 0) {
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ResourceInUserService.java]: Have you checked for newQuantity = 0.00? I remember there was some weird behaviour comparing double to 0.

File: src/test/java/jewellery/inventory/integration/ResourceInUserCrudIntegrationTest.java [hunk 22]
```diff
@@ -75,28 +78,31 @@ void setup() {
...
 
     ResourceInUserRequestDto requestDto =
         createResourceInUserRequestDto(
             createdUser.getId(), createdResource.getId(), RESOURCE_QUANTITY);
 
-    ResponseEntity<UserResponseDto> response =
-        testRestTemplate.postForEntity(
-            getBaseResourceAvailabilityUrl(), requestDto, UserResponseDto.class);
+    System.out.println(createdUser.getId());
+    System.out.println(createdResource.getId());
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/ResourceInUserCrudIntegrationTest.java]: let's not leave system.out in the code pls

File: src/main/java/jewellery/inventory/dto/response/resource/ResourceInUserResponseDto.java [hunk 23]
```diff
@@ -1,12 +1,12 @@
...
+import jewellery.inventory.dto.response.UserResponseDto;
 import lombok.Data;
 
 @Data
 public class ResourceInUserResponseDto {
-  private UUID resourceId;
-  private String ownerName;
-  private String resourceClazz;
+  private UserResponseDto owner;
+  List<ResourceResponseDto> resources;
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/dto/response/resource/ResourceInUserResponseDto.java]: The list should be a list of (resource, quantity) <=> List<Pair<ResourceResponseDto, Double>>

**Review by VladoKat — CHANGES_REQUESTED** [review]
Please, take a look at the comments. There is one major thing that needs to be changed - the ResourceInUserResponseDto. It could be renamed to Resource**s**InUserResponseDto.

Another thing that is not mentioned but we might want to do is to make the ResourceInUser<smth> to be outside the resource packages in the model and dto packages. It feels a bit out of place there as the other are direct descendants of Resource.

File: src/main/java/jewellery/inventory/dto/response/resource/ResourceInUserResponseDto.java (hunk 23 above)
Comment (VladoKat) [reply src/main/java/jewellery/inventory/dto/response/resource/ResourceInUserResponseDto.java]: ResourceInUserRequest should remain as it is.

File: src/main/java/jewellery/inventory/exception/not_found/ResourceInUserNotFoundException.java (hunk 19 above)
Comment (VladoKat) [reply src/main/java/jewellery/inventory/exception/not_found/ResourceInUserNotFoundException.java]: What is the difference between this and `UserDoesNotOwnResourceException`?

File: src/main/java/jewellery/inventory/service/ResourceInUserService.java [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+        findResourceInUser(user, resourceUserDto.getResourceId())
+            .orElseGet(() -> createAndAddNewResourceInUser(user, resource, 0));
+
+    resourceInUser.setQuantity(resourceInUser.getQuantity() + resourceUserDto.getQuantity());
+
+    return UserMapper.INSTANCE.toResourceInUserResponseDto(
+        resourceInUserRepository.save(resourceInUser));
+  }
+
+  public List<ResourceInUserResponseDto> getAllResourcesFromUser(UUID userId) {
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ResourceInUserService.java]: After the ResourcesInUserResponseDto, it would make this gets all the ResourceInUser owned by a user and returns a single ResourcesInUserResponseDto

File: src/test/java/jewellery/inventory/integration/ResourceInUserCrudIntegrationTest.java [hunk 25]
```diff
@@ -0,0 +1,386 @@
...
+    return "http://localhost:" + port + "/resources/availability/" + userId + "/" + resourceId;
+  }
+
+  private String getBaseUserUrl() {
+    return "http://localhost:" + port + "/users";
+  }
+
+  private String getBaseResourceUrl() {
+    return "http://localhost:" + port + "/resources";
+  }
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/ResourceInUserCrudIntegrationTest.java]: Lots of hardcoded strings repetitions. Could we extract at least getBaseUrl() to return "http://localhost:" + port

File: src/test/java/jewellery/inventory/integration/ResourceInUserCrudIntegrationTest.java (hunk 22 above)
Comment (Alekseeybg) [reply src/test/java/jewellery/inventory/integration/ResourceInUserCrudIntegrationTest.java]: Fixed

File: src/main/java/jewellery/inventory/exception/GlobalExceptionHandler.java (hunk 17 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/exception/GlobalExceptionHandler.java]: Fixed

File: src/main/java/jewellery/inventory/exception/GlobalExceptionHandler.java (hunk 18 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/exception/GlobalExceptionHandler.java]: Fixed

File: src/test/java/jewellery/inventory/integration/ResourceInUserCrudIntegrationTest.java (hunk 25 above)
Comment (Alekseeybg) [reply src/test/java/jewellery/inventory/integration/ResourceInUserCrudIntegrationTest.java]: Fixed

File: src/main/java/jewellery/inventory/dto/response/resource/ResourceInUserResponseDto.java (hunk 23 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/dto/response/resource/ResourceInUserResponseDto.java]: Created a separate DTO to fix the issue. I believe this is the correct approach in this case.

File: src/main/java/jewellery/inventory/exception/not_found/ResourceInUserNotFoundException.java (hunk 19 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/exception/not_found/ResourceInUserNotFoundException.java]: Removed UserDoesNotOwnResourceException. I really do not know why I basically duplicated it... I was thinking of including the user UUID as well but did not want the error message to be too long. Fixed.

File: src/main/java/jewellery/inventory/exception/GlobalExceptionHandler.java (hunk 17 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/exception/GlobalExceptionHandler.java]: It was a duplicate of ResourceInUserNotFoundException. Deleted.

File: src/main/java/jewellery/inventory/dto/response/resource/ResourceInUserResponseDto.java (hunk 23 above)
Comment (VladoKat) [reply src/main/java/jewellery/inventory/dto/response/resource/ResourceInUserResponseDto.java]: What is the scenario in which we are going to use this response? The list of resource dtos can't map to a single quantity field.

File: src/main/java/jewellery/inventory/service/ResourceInUserService.java (hunk 24 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/service/ResourceInUserService.java]: Fixed.

File: src/main/java/jewellery/inventory/service/ResourceInUserService.java (hunk 21 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/service/ResourceInUserService.java]: I did not know about that issue! Fixed.

File: src/main/java/jewellery/inventory/mapper/UserMapper.java (hunk 20 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/mapper/UserMapper.java]: I think you were correct. Not sure if the refactored code fixed this issue though.

File: src/main/java/jewellery/inventory/dto/response/resource/ResourceInUserResponseDto.java (hunk 23 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/dto/response/resource/ResourceInUserResponseDto.java]: At the time you wrote your comment the new code was not pushed yet. The new DTO is not a response DTO. It is used in the ResourcesInUserResponseDto. The new version is
```@Data
public class ResourcesInUserResponseDto {
  private UserResponseDto owner;
  private List<ResourceQuantityDto> resources;
}
```
And the new DTO looks like this:
```@Data
public class ResourceQuantityDto {
    private ResourceResponseDto resource;
    private double quantity;
}
```
Maybe the naming ResourceQuantityDto  is not the best

Comment (VladoKat) [reply src/main/java/jewellery/inventory/dto/response/resource/ResourceInUserResponseDto.java]: yep, this looks better. I don't mind the name. Maybe more appropriate would be ResourceAndQuantityDto

File: src/main/java/jewellery/inventory/service/ResourceInUserService.java [hunk 26]
```diff
@@ -40,15 +43,22 @@ public ResourceInUserResponseDto addResourceToUser(ResourceInUserRequestDto reso
...
+    List<ResourceQuantityDto> resourcesWithQuantities =
+        user.getResourcesOwned().stream()
+            .map(resourcesInUserMapper::toResourceQuantityDto)
+            .collect(Collectors.toList());
+
+    ResourcesInUserResponseDto responseDto = new ResourcesInUserResponseDto();
+    responseDto.setOwner(resourcesInUserMapper.toUserResponse(user));
+    responseDto.setResources(resourcesWithQuantities);
+
+    return responseDto;
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ResourceInUserService.java:61]: Can we make a mapping method in.. (UserMapper?) that maps User (entity) to ResourcesInUserResponseDto?

**Review by VladoKat — COMMENTED** [review]
One minor comment. Would approve after we discuss it.

File: src/main/java/jewellery/inventory/service/ResourceInUserService.java (hunk 26 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/service/ResourceInUserService.java:61]: Added it to ResourcesInUserMapper

**Comment by VladoKat** [comment]
duplicate of https://github.com/JewelleryManagement/jewellery-management-service/pull/23
